### ISSUE TITLE
Unexpected “Spread types may only be created from object types” error when using generics

### ISSUE THREAD

**Issue by Olian04** [Microsoft/TypeScript#17368]
**TypeScript Version:**  2.4.2

**Code**

```ts
type Partial<T> = {
  [P in keyof T]?: T[P];
};

class Foo<Bar> {
  bis: Partial<Bar> = {}; // (1)
  constructor() {
    console.log(typeof this.bis);  // object
    this.bis = {...this.bis};  // (2) Spread types may only be created from object types
  }
}
```

**Expected behavior:**
Either allow the spreading at _(2)_ or throw an error at _(1)_.

**Actual behavior:**
_(1)_ doesn't throw an error, however _(2)_ does.

**Comment by jcalz** [comment]
#10727

**Comment by Olian04** [comment]
@jcalz I don't think so? The spread operator is already part of ts, and what their proposing looks more like an addition to the spread operator. This issue is meant to show a possible bug in the current implementation of spread where it fails in interact with the type system.

**Comment by jcalz** [comment]
See discussion in #12756

**Comment by Olian04** [comment]
@jcalz Still not convinced its the same.

**Comment by mhegazy** [comment]
`this` in a class is a generic type parameter, it is the type of the most specialized class that may extend from this class.
https://github.com/Microsoft/TypeScript/issues/10727 tracks enabling the spread operator on generic types.

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.
